docs/source/extending_pywr/extending_pywr_parameters.rst

```diff
@@ -0,0 +1,224 @@
+.. _extending-pywr-parameters:
+
+Extending Pywr with custom Parameters
+-------------------------------------
+
+Pywr provides many existing Parameters that can be used to model complex behaviours. This includes parameters that
+read data in different formats (e.g. the dataframe parameter), parameters that solve specific problems (e.g. reservoir
+control curves) and general purpose parameters (e.g. the ``AggregatedParameter``).
+
+Custom parameters can be used to model specific behaviours otherwise not possible with the existing parameters. To
+write a custom parameter you must (as a minimim) inherit from the base ``pywr.parameter.Parameter`` class and implement
+the ``Parameter.value`` method. The arguments to this method represent the current timestep and scenario allowing the
+value of the parameter to be dynamic.
+
+A simple parameter
+==================
+
+The example below shows a minimal parameter which returns the same value every timestep. The parameter is initialised
+with the value to return, which it stores in the ``_value`` attribute on the instance [*]_. This value is returned
+whenever the ``value`` method is queries. The ``load`` class method is used to create an instance of the parameter
+from JSON.
+
+.. code-block:: python
+
+    from pywr.parameters import Parameter
+
+    class MyParameter(Parameter):
+        def __init__(self, model, value):
+            # called once when the parameter is created
+            super().__init__(model)
+            self._value = value
+
+        def value(self, timestep, scenario_index):
+            # called once per timestep for each scenario
+            return self._value
+
+        @classmethod
+        def load(cls, model, data):
+            # called when the parameter is loaded from a JSON document
+            value = data.pop("value")
+            return cls(model, value)
+
+    MyParameter.register()  # register the name so it can be loaded from JSON
+
+.. [*] ``_value`` is used instead of ``value`` to avoid overloading the method of the same name.
+
+An instance of the parameter can be created from a JSON model as below. The type of the parameter is the class name,
+with the word "parameter" optionally removed (e.g. ``"constant"`` and ``"constantparameter"`` both create an instance
+of ``ConstantParameter``).
+
+.. code-block:: yaml
+
+    "max_flow": {
+        "type": "myparameter",
+        "value": 123.0
+    }
+
+Timesteps and scenarios
+=======================
+
+The ``Parameter.value`` method is called once per timestep for each scenario. The value it returns can be varied using
+the ``timestep`` and ``scenario_index`` arguments. For example, a simple version of a ``MonthlyProfileParameter`` could be
```

Comment:
Are the `Timestep` and `ScenarioIndex` documented yet? It would be good to point out that these arguments are pywr types. One might think they are a `datetime` and `int`.

Reply:
We should link them to the autodoc for those classes. I can't remember how to do this in rst.

Reply:
Created an issue for adding `pywr.core` to the autodoc, which includes `Timestep` and `ScenarioIndex`. #689